## One-day International (ODI) Cricket Data Analysis of Bowlers

Author: Mahsab Al Rahman

Objective:
The goal of this notebook is to analyze one-day international (ODI) cricket data of different bowlers from the given dataset


### Task 1: Import required libraries

In [361]:
import pandas as pd
import numpy as np

### Reading the excel file

In [362]:
df = pd.read_excel("ODI_cricket.xlsx", sheet_name="bowler", engine="openpyxl")

display(df.head())

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (Asia/ICC/SL)  1993-2011  350   341  18811  12326   534   
1             Wasim Akram (PAK)  1984-2003  356   351  18186  11812   502   
2            Waqar Younis (PAK)  1989-2003  262   258  12698   9919   416   
3          WPUJC Vaas (Asia/SL)  1994-2008  322   320  15775  11014   400   
4  Shahid Afridi (Asia/ICC/PAK)  1996-2015  398   372  17670  13632   395   

     Ave  Econ    SR   4   5  
0  23.08  3.93  35.2  15  10  
1  23.52  3.89  36.2  17   6  
2  23.84  4.68  30.5  14  13  
3  27.53  4.18  39.4   9   4  
4  34.51  4.62  44.7   4   9

### Task 2: Display the first 10 rows of the dataframe

In [363]:
df = pd.read_excel("ODI_cricket.xlsx", sheet_name="bowler", engine="openpyxl")

display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (Asia/ICC/SL)  1993-2011  350   341  18811  12326   534   
1             Wasim Akram (PAK)  1984-2003  356   351  18186  11812   502   
2            Waqar Younis (PAK)  1989-2003  262   258  12698   9919   416   
3          WPUJC Vaas (Asia/SL)  1994-2008  322   320  15775  11014   400   
4  Shahid Afridi (Asia/ICC/PAK)  1996-2015  398   372  17670  13632   395   
5       SM Pollock (Afr/ICC/SA)  1996-2008  303   297  15712   9631   393   
6          GD McGrath (AUS/ICC)  1993-2007  250   248  12970   8391   381   
7                   B Lee (AUS)  2000-2012  221   217  11185   8877   380   
8               SL Malinga (SL)  2004-2019  226   220  10936   9760   338   
9         A Kumble (Asia/INDIA)  1990-2007  271   265  14496  10412   337   

     Ave  Econ    SR   4   5  
0  23.08  3.93  35.2  15  10  
1  23.52  3.89  36.2  17   6  
2  23.84  4.68  30.5  14  13  
3  27.53  4.18  39.4   9   4  
4  34.51  4.62  44.7   4   9  
5  24.50  3.67  39.9  12   5  
6  22.02  3.88  34.0   9   7  
7  23.36  4.76  29.4  14   9  
8  28.87  5.35  32.3  11   8  
9  30.89  4.30  43.0   8   2

### Task 3: Create a markdown cell and explain the meaning of each column

Player: Name of the players who played ODI cricket

Span: Career span of each player

Mat: Number of matches played

Inns: Total innings played by the player

Balls: Total number of balls bowled by the player

Runs: The number of runs conceded by the player

Wkts: Total number of wickets taken by the player

Ave: The average number of runs conceded for per wicket by the player (Runs/Wicket)

Econ: The average number of runs conceded per over (Runs/Over)

SR: The average number of balls bowled per wicket taken (Balls/Wicket)

5: The number of innings in which the bowler took at least five wickets

10: The number of matches in which the bowler took at least ten wickets

### Task 4: Find the number of rows and columns in the dataframe

In [364]:
# number of rows
print("number of rows = ", df.shape[0])

# number of columns
print("number of columns = ", df.shape[1])

number of rows =  77
number of columns =  12


### Task 5:  Find the data statistics and check for the data types

In [365]:
#checking data types
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  77 non-null     object 
 1   Span    77 non-null     object 
 2   Mat     77 non-null     int64  
 3   Inns    77 non-null     int64  
 4   Balls   77 non-null     int64  
 5   Runs    77 non-null     int64  
 6   Wkts    77 non-null     int64  
 7   Ave     77 non-null     float64
 8   Econ    77 non-null     float64
 9   SR      77 non-null     float64
 10  4       77 non-null     int64  
 11  5       77 non-null     int64  
dtypes: float64(3), int64(7), object(2)
memory usage: 7.3+ KB


None

Observations:

There are

3 float type variables: Bowling average, Economy rate, Strike rate

2 String type variables: Player, Span

7 integer type variables: Match, Innings, Balls, Runs, Wickets, 5 wickets in and innings, 10 wickets in an innings


From the non-null columns, we can see that there are no missing value as of now.

In [366]:
#Showing data statistics
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   77.000000   77.000000     77.000000     77.000000   77.000000   
mean   194.402597  181.194805   8839.402597   6671.714286  233.805195   
std     82.485606   67.958393   3316.055457   2245.839029   84.406603   
min     80.000000   76.000000   4074.000000   2821.000000  151.000000   
25%    136.000000  128.000000   6182.000000   5058.000000  173.000000   
50%    170.000000  164.000000   8054.000000   6192.000000  199.000000   
75%    227.000000  218.000000  10750.000000   8021.000000  272.000000   
max    463.000000  372.000000  18811.000000  13632.000000  534.000000   

             Ave       Econ         SR          4          5  
count  77.000000  77.000000  77.000000  77.000000  77.000000  
mean   28.958052   4.596753  37.909091   6.350649   2.870130  
std     4.826768   0.515814   6.060901   3.556929   2.530606  
min    18.680000   3.300000  26.100000   1.000000   0.000000  
25%    24.970000   4.280000  33.000000   3.000000   1.000000  
50%    29.290000   4.660000  37.800000   6.000000   2.000000  
75%    31.900000   4.920000  41.400000   8.000000   4.000000  
max    44.480000   5.830000  52.500000  17.000000  13.000000

Statistical Measures:

Count: total numbers of observation


mean: average of the values

std: standard deviation

min: minimum value in the column

25%: 1st quartile or 25% of all value is under the certain column value

50%: 2nd quartile or 50% of all value is under the certain column value

75%: 3rd quartie or 75% of all value is under the certain column value

max: maximum value in the column


Observations:

When it comes to innings, it is observed that:

On an average a played has played around 181 innings.

25% of the players have played 128 innings or less

50% of the the players have played 164 innings or less

75% of the players have played 218 innings or less

Maximum 372 innings were played by a player

Minimum 76 innings were played by a player

Many other insights can be taken from this statistics.


### Task 6: Are there any missing values present in the dataset?

While checking for the data types in this file, it was found out that there is no missing value.

### Task 7: Rename the column names appropriately

In [367]:
# renaming columns in dataframe
df = df.rename(columns={'Span' : 'Career_Duration',
                        'Mat' : 'Matches',
                        'Inns': 'Innings',
                        'Wkts' : 'Wickets',
                        'Ave' : 'Bowling_Average',
                        'Econ': 'Economy_Rate',
                        'SR' : 'Strike_Rate',
                        4: '4_Wickets in an innings',
                        5: '5_Wickets in an innings'})
display(df.head(10))

Player Career_Duration  Matches  Innings  Balls  \
0  M Muralitharan (Asia/ICC/SL)       1993-2011      350      341  18811   
1             Wasim Akram (PAK)       1984-2003      356      351  18186   
2            Waqar Younis (PAK)       1989-2003      262      258  12698   
3          WPUJC Vaas (Asia/SL)       1994-2008      322      320  15775   
4  Shahid Afridi (Asia/ICC/PAK)       1996-2015      398      372  17670   
5       SM Pollock (Afr/ICC/SA)       1996-2008      303      297  15712   
6          GD McGrath (AUS/ICC)       1993-2007      250      248  12970   
7                   B Lee (AUS)       2000-2012      221      217  11185   
8               SL Malinga (SL)       2004-2019      226      220  10936   
9         A Kumble (Asia/INDIA)       1990-2007      271      265  14496   

    Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
0  12326      534            23.08          3.93         35.2   
1  11812      502            23.52          3.89         36.2   
2   9919      416            23.84          4.68         30.5   
3  11014      400            27.53          4.18         39.4   
4  13632      395            34.51          4.62         44.7   
5   9631      393            24.50          3.67         39.9   
6   8391      381            22.02          3.88         34.0   
7   8877      380            23.36          4.76         29.4   
8   9760      338            28.87          5.35         32.3   
9  10412      337            30.89          4.30         43.0   

   4_Wickets in an innings  5_Wickets in an innings  
0                       15                       10  
1                       17                        6  
2                       14                       13  
3                        9                        4  
4                        4                        9  
5                       12                        5  
6                        9                        7  
7                       14                        9  
8                       11                        8  
9                        8                        2

### Task 8: How many players played for ICC?

In [368]:
# removing the closing bracket
df['Player'] = df['Player'].str.replace(")"," ")  

# splitting
df_player = df['Player'].str.split("(", expand=True)

# dropping old player column
df = df.drop('Player', axis=1)  

# concatenating the df_player with main dataframe
df = pd.concat([df, df_player], axis=1) 

# renaming the columns
df = df.rename(columns={0: 'Player',1: 'Country'})  

# rearrange the columns
new_col_sequence = ['Player', 'Country', 'Career_Duration', 'Matches', 'Innings',
                    'Wickets', 'Bowling_Average', 'Economy_Rate', 'Strike_Rate', 'Balls', 'Runs', '4_Wickets in an innings', '5_Wickets in an innings']

df = df[new_col_sequence]

display(df.head(10))

C:\Users\Lenovo\AppData\Local\Temp/ipykernel_8872/123548211.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['Player'] = df['Player'].str.replace(")"," ")


Player        Country Career_Duration  Matches  Innings  Wickets  \
0  M Muralitharan    Asia/ICC/SL        1993-2011      350      341      534   
1     Wasim Akram            PAK        1984-2003      356      351      502   
2    Waqar Younis            PAK        1989-2003      262      258      416   
3      WPUJC Vaas        Asia/SL        1994-2008      322      320      400   
4   Shahid Afridi   Asia/ICC/PAK        1996-2015      398      372      395   
5      SM Pollock     Afr/ICC/SA        1996-2008      303      297      393   
6      GD McGrath        AUS/ICC        1993-2007      250      248      381   
7           B Lee            AUS        2000-2012      221      217      380   
8      SL Malinga             SL        2004-2019      226      220      338   
9        A Kumble     Asia/INDIA        1990-2007      271      265      337   

   Bowling_Average  Economy_Rate  Strike_Rate  Balls   Runs  \
0            23.08          3.93         35.2  18811  12326   
1            23.52          3.89         36.2  18186  11812   
2            23.84          4.68         30.5  12698   9919   
3            27.53          4.18         39.4  15775  11014   
4            34.51          4.62         44.7  17670  13632   
5            24.50          3.67         39.9  15712   9631   
6            22.02          3.88         34.0  12970   8391   
7            23.36          4.76         29.4  11185   8877   
8            28.87          5.35         32.3  10936   9760   
9            30.89          4.30         43.0  14496  10412   

   4_Wickets in an innings  5_Wickets in an innings  
0                       15                       10  
1                       17                        6  
2                       14                       13  
3                        9                        4  
4                        4                        9  
5                       12                        5  
6                        9                        7  
7                       14                        9  
8                       11                        8  
9                        8                        2

In [369]:
 # filter function 
def ICC(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"
    
def asia_check(x):
    if "Asia" in x:
        return "Yes"
    else:
        return "No"
    
def africa_check(x):
    if "Afr" in x:
        return "Yes"
    else:
        return "No"

# using filter function to check
df['Played_for_ICC'] = df['Country'].apply(ICC)
df['Played_for_Asia'] = df['Country'].apply(asia_check)
df['Played_for_Africa'] = df['Country'].apply(africa_check)

# removing "ICC/" from string of country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
df['Country'] = df['Country'].str.replace("/ICC","", regex=True)
df['Country'] = df['Country'].str.replace("Asia/", "", regex=True)
df['Country'] = df['Country'].str.replace("Afr/", "", regex=True)

display(df.head(5))

Player Country Career_Duration  Matches  Innings  Wickets  \
0  M Muralitharan      SL        1993-2011      350      341      534   
1     Wasim Akram     PAK        1984-2003      356      351      502   
2    Waqar Younis     PAK        1989-2003      262      258      416   
3      WPUJC Vaas      SL        1994-2008      322      320      400   
4   Shahid Afridi     PAK        1996-2015      398      372      395   

   Bowling_Average  Economy_Rate  Strike_Rate  Balls   Runs  \
0            23.08          3.93         35.2  18811  12326   
1            23.52          3.89         36.2  18186  11812   
2            23.84          4.68         30.5  12698   9919   
3            27.53          4.18         39.4  15775  11014   
4            34.51          4.62         44.7  17670  13632   

   4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
0                       15                       10            Yes   
1                       17                        6             No   
2                       14                       13             No   
3                        9                        4             No   
4                        4                        9            Yes   

  Played_for_Asia Played_for_Africa  
0             Yes                No  
1              No                No  
2              No                No  
3             Yes                No  
4             Yes                No

In [370]:
# checking count of player who played for ICC
print(df['Played_for_ICC'].value_counts())

No     64
Yes    13
Name: Played_for_ICC, dtype: int64


Observatiuon:
13 players played for ICC

### Task 9: How many different countries are present in this dataset? 

In [371]:
# no. of countries
number_of_different_countries=len(df['Country'].unique()) 

print(number_of_different_countries, "different countries are present in the dataset")

11 different countries are present in the dataset


### Task 10: Which player(s) had played for the longest period of time?

In [372]:
# splitting 'Career_Duration' column
df_career_duration = df['Career_Duration'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_career_duration], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Start_Year",
                        1: "End_Year"})

# removing 'Career_Duration' column
df = df.drop("Career_Duration", axis=1)


In [373]:
# Changing the data type of Start and End Year from object to integer
df['End_Year']=pd.to_numeric(df['End_Year'])
df['Start_Year']=pd.to_numeric(df['Start_Year'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Player                   77 non-null     object 
 1   Country                  77 non-null     object 
 2   Matches                  77 non-null     int64  
 3   Innings                  77 non-null     int64  
 4   Wickets                  77 non-null     int64  
 5   Bowling_Average          77 non-null     float64
 6   Economy_Rate             77 non-null     float64
 7   Strike_Rate              77 non-null     float64
 8   Balls                    77 non-null     int64  
 9   Runs                     77 non-null     int64  
 10  4_Wickets in an innings  77 non-null     int64  
 11  5_Wickets in an innings  77 non-null     int64  
 12  Played_for_ICC           77 non-null     object 
 13  Played_for_Asia          77 non-null     object 
 14  Played_for_Africa        77 

In [374]:
# Calculating career span from start and end year

# total years of career of the players
df['Career_Span']=(df['End_Year']-df['Start_Year']) + 1

# new column sequence as a list
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Career_Span', 'Matches', 'Innings',
                    'Wickets', 'Bowling_Average', 'Economy_Rate', 'Strike_Rate', 'Balls', 'Runs', '4_Wickets in an innings', '5_Wickets in an innings', 'Played_for_ICC', 'Played_for_Asia', 'Played_for_Africa']

# implementing the new column sequence
df = df[column_sequence]

display(df.head())

Player Country  Start_Year  End_Year  Career_Span  Matches  \
0  M Muralitharan      SL         1993      2011           19      350   
1     Wasim Akram     PAK         1984      2003           20      356   
2    Waqar Younis     PAK         1989      2003           15      262   
3      WPUJC Vaas      SL         1994      2008           15      322   
4   Shahid Afridi     PAK         1996      2015           20      398   

   Innings  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls   Runs  \
0      341      534            23.08          3.93         35.2  18811  12326   
1      351      502            23.52          3.89         36.2  18186  11812   
2      258      416            23.84          4.68         30.5  12698   9919   
3      320      400            27.53          4.18         39.4  15775  11014   
4      372      395            34.51          4.62         44.7  17670  13632   

   4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
0                       15                       10            Yes   
1                       17                        6             No   
2                       14                       13             No   
3                        9                        4             No   
4                        4                        9            Yes   

  Played_for_Asia Played_for_Africa  
0             Yes                No  
1              No                No  
2              No                No  
3             Yes                No  
4             Yes                No

In [375]:
# Which player(s) had played for the longest period of time?
display(df.sort_values(by='Career_Span', ascending = False).head(5))

Player Country  Start_Year  End_Year  Career_Span  Matches  \
72   SR Tendulkar   INDIA         1989      2012           24      463   
10  ST Jayasuriya      SL         1989      2011           23      445   
66   Shoaib Malik     PAK         1999      2019           21      287   
62       CH Gayle      WI         1999      2019           21      301   
4   Shahid Afridi     PAK         1996      2015           20      398   

    Innings  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls  \
72      270      154            44.48          5.10         52.2   8054   
10      368      323            36.75          4.78         46.0  14874   
66      217      158            39.18          4.66         50.3   7958   
62      199      167            35.48          4.78         44.4   7424   
4       372      395            34.51          4.62         44.7  17670   

     Runs  4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
72   6850                        4                        2             No   
10  11871                        8                        4             No   
66   6192                        1                        0             No   
62   5926                        3                        1            Yes   
4   13632                        4                        9            Yes   

   Played_for_Asia Played_for_Africa  
72              No                No  
10             Yes                No  
66              No                No  
62              No                No  
4              Yes                No

Observation:
SR Tendulkar had the longest career span of about 24 years.

### Task 11: Which player(s) had played for the shortest period of time?

In [376]:
display(df.sort_values(by='Career_Span', ascending = True).head(5))

Player Country  Start_Year  End_Year  Career_Span  Matches  Innings  \
40   BKV Prasad   INDIA         1994      2001            8      161      160   
74  Rashid Khan     AFG         2015      2022            8       80       76   
73   BAW Mendis      SL         2008      2015            8       87       84   
49  Saeed Ajmal     PAK         2008      2015            8      113      112   
44   L Klusener      SA         1996      2004            9      171      164   

    Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls  Runs  \
40      196            32.30          4.67         41.4   8129  6332   
74      151            18.68          4.15         26.9   4074  2821   
73      152            21.86          4.80         27.3   4154  3324   
49      184            22.72          4.18         32.6   6000  4182   
44      192            29.95          4.70         38.2   7336  5751   

    4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
40                        3                        1             No   
74                        5                        4             No   
73                        7                        3             No   
49                        6                        2             No   
44                        1                        6             No   

   Played_for_Asia Played_for_Africa  
40              No                No  
74              No                No  
73              No                No  
49              No                No  
44              No                No

Observation:
BKV Prasad had the shortest career span of abround 8 years.

### Task 12: How many Australian Bowlers are present in this dataset?

In [377]:
print(df['Country'].value_counts())

INDIA     13
PAK       12
AUS       10
SL         9
SA         8
NZ         8
ENG        6
WI         6
BAN        3
ZIM        1
AFG        1
Name: Country, dtype: int64


10 bowlers from Australia are present in this dataset.

### Task 13: Is there any Bangladeshi player present in this dataset?

Yes. 3 Bowlers.

### Task 14: Which player had the lowest economy rate?

In [378]:
display(df.sort_values(by='Economy_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Matches  Innings  \
65  RJ Hadlee      NZ         1973      1990           18      115      112   

    Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls  Runs  \
65      158            21.56           3.3         39.1   6182  3407   

    4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
65                        1                        5             No   

   Played_for_Asia Played_for_Africa  
65              No                No

Observation:
RJ Hadlee had the lowest economy rate of about 3.3 runs per over.

### Task 15: Which player had the lowest strike rate?

In [379]:
display(df.sort_values(by='Strike_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Matches  Innings  \
41  MA Starc     AUS         2010      2021           12       99       99   

    Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls  Runs  \
41      195            22.45          5.15         26.1   5099  4379   

    4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
41                       11                        8             No   

   Played_for_Asia Played_for_Africa  
41              No                No

Observation:
MA Starc has the lowest srtike rate of about 26.1 balls per wicket.

### Task 16: Which player had the lowest bowling average?

In [380]:
display(df.sort_values(by='Bowling_Average', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Matches  Innings  \
74  Rashid Khan     AFG         2015      2022            8       80       76   

    Wickets  Bowling_Average  Economy_Rate  Strike_Rate  Balls  Runs  \
74      151            18.68          4.15         26.9   4074  2821   

    4_Wickets in an innings  5_Wickets in an innings Played_for_ICC  \
74                        5                        4             No   

   Played_for_Asia Played_for_Africa  
74              No                No

Observation:
Rashid Khan has the lowest bowling average of around 18.68 runs per wicket.

### Task 17: Remove Unnecessary columns if needed

In my opinion, there are no columns here that can be treated as redundant.